urllib3/connection.py

```diff
@@ -299,7 +302,18 @@ def connect(self):
                     'for details.)'.format(hostname)),
                     SubjectAltNameWarning
                 )
-            match_hostname(cert, self.assert_hostname or hostname)
+            asserted_hostname = self.assert_hostname or hostname
+            try:
+                match_hostname(cert, asserted_hostname)
+            except CertificateError as e:
+                log.error(
+                    'Certificate did not match expected hostname: %s. '
+                    'Certificate: %s', asserted_hostname, cert
+                )
+                # Add cert to exception and reraise so client code can inspect
+                # the cert when catching the exception, if they want to
+                e.cert = cert
```

Comment:
Should we be wrapping this in a urllib3-local exception that always has this field to avoid monkeypatching?